### ISSUE TITLE
Allow passing in properties of components as inputs

### ISSUE THREAD

**Issue by abidlabs** [gradio-app/gradio#5220]
We frequently get asked how to pass in properties of components as inputs (kind of the "opposite" of returning a `gr.update()`). For example, see:

* https://github.com/gradio-app/gradio/issues/3793
* Internal: https://huggingface.slack.com/archives/C02QZLG8GMN/p1692028185129229

We don't currently support this and our general answer to is to use the component event trigger + store the state in a `gr.State`, which is then used as the input. However, this approach is quite cumbersome. We should find a way to make this easier.

Idea 1:

* When a component is used as an input, we should pass in a custom class whose primary value is the value of the component, but which also has attributes that reflect the attributes of the property. For example:

```py
def get_tab_index(tabs):
  return tabs.selected

with gr.Blocks() as demo:
  with gr.Tabs() as tab:
  ...
  btn.click(get_tab_index, tabs, number)
```

However, the problem with this is is that it would not be compatible with using Gradio apps via an API.

Idea 2:

* You can pass in a component prop as an input. For example:

```py
with gr.Blocks() as demo:
  with gr.Tabs() as tab:
  ...
  btn.click(lambda x:x, tabs.selected, number)
```

We should be able to make this work with our API (the typing info would be inferred from the upcoming component dataclass in 4.0).

**Comment by freddyaboulton** [comment]
I like idea 1 and was brought up [here](https://github.com/gradio-app/gradio/issues/4926) but in addition to complicating the api usage, it couples the function definition with the gradio app state.

**Comment by pngwn** [comment]
The only issue with this is it may lead to very large payloads that are totally unused most of the time. Unless we can either figure out exactly what is being used when generating the config, or we implement the patching that @aliabid94 mentioned a little while ago in _both_ directions.

**Comment by elismasilva** [comment]
hi @abidlabs any new about this ? 
I have a situation, i am using dynamic app rendering components and i am trying to change a value of one component and recover this value on my submit function. 

![image](https://github.com/user-attachments/assets/ea4ddfe8-98f1-4882-a6f5-127f607f1506)

so i have this event function that is trigged by change event of lora scale component.
``` python
def change_lora_scale(value):
    lora_scale_idx = loras_scales.index(value)
    loras_scales[lora_scale_idx]=value
```
the value is 1.2 like in image, but iam trying recover the component that ive added previously on render block like this:

``` python
lora = gr.Textbox(label="File name", value=f"{selected[i]}", key=f"label-{selected[i]}", show_label=True) 
....
if lora_scale not in loras_scales:
    loras_scales.append(lora_scale)
```
So i need receive the key propertie that i defined in textbox where i can do a for loop on lora_scales to recover compoment by key and then change the value.

why couldn't event functions just receive a kwargs property so we can put whatever we want in it?

**Comment by elismasilva** [comment]
sorry because i am very nubie on gradio lol, but i solve my problem doing this:

i moved change_lora_scale from outside block to inside render block where the dynamics components are created:

``` python
 @gr.render(inputs=selected_loras_state)
  def render_loras(selected):                                                       
      with gr.Row(elem_id="lora_row"):                                                        
          for i in range(len(selected)):                                                                                                                        
              with gr.Column(variant="panel", min_width=300):
...
                    def change_lora_scale(value):
                    for l in range(len(loras_scales)):
                        if loras_scales[l].key[8:] == selected[i]:
                            loras_scales[l].value=value             
```
Since "selected" already have my key value i can use it inside block to recover my previous saved component.

**Comment by elismasilva** [comment]
but gradio could be have a method like in DOM to recover components by ID or key like;

```python
with gr.Blocks() as app
     .... do components...
component = app.getElementByID(id)
```
so if i can do it from everywhere i dont need add components in a list to recover after...

**Comment by abidlabs** [comment]
Idea 3: pass in a special attribute of a component, which passes in the entire "config" of that component

```py
with gr.Blocks() as demo:
  with gr.Tabs() as tab:
  ...
  btn.click(lambda x:x, tabs.props, number)
```

**Comment by acylam** [comment]
> but gradio could be have a method like in DOM to recover components by ID or key like;
> 
> ```python
> with gr.Blocks() as app
>      .... do components...
> component = app.getElementByID(id)
> ```
> 
> so if i can do it from everywhere i dont need add components in a list to recover after...

This is one thing I was looking for since I started using gradio. Needing to pass a list of inputs and outputs seem a bit too cumbersome for large applications.

**Comment by abidlabs** [comment]
> This is one thing I was looking for since I started using gradio. Needing to pass a list of inputs and outputs seem a bit too cumbersome for large applications. Especially when the list is long, it's hard to track the number of output components and their order.

That's fair, we'll brainstorm around this

**Comment by morisono** [comment]
Any new idea about this? This is important for complex tasks solving.
[See the example code on Playground](https://www.gradio.app/playground?demo=Hello_World&code=aW1wb3J0IGdyYWRpbyBhcyBncg0KDQpkZWYgc2VuZF9hbmRfc3dpdGNoKHN0YXQsIG4pOg0KICAgIGdyLkluZm8oZiJDaGFuZ2luZyB0YWIgdG8ge259IikNCiAgICByZXR1cm4gZ3IuVGFicyhzZWxlY3RlZD1uKSwgaW5wDQoNCmRlZiBpb19jb21wb25lbnRzKHN0YXQpOg0KICAgIHdpdGggZ3IuUm93KCk6DQogICAgICAgIHdpdGggZ3IuQ29sdW1uKCk6DQogICAgICAgICAgICBpbnAgPSBnci5NdWx0aW1vZGFsVGV4dGJveChsYWJlbD0iRW50ZXIgdGV4dCIpDQogICAgICAgIHdpdGggZ3IuQ29sdW1uKCk6DQogICAgICAgICAgICBuID0gZ3IuRHJvcGRvd24oY2hvaWNlcz1saXN0KHN0YXQudmFsdWVbInRhYnMiXS5rZXlzKCkpLCBpbnRlcmFjdGl2ZT1UcnVlLCB2YWx1ZT1zdGF0LnZhbHVlWyJ0YWIiXSwgbGFiZWw9IlNlbmQgdG8uLiIpDQogICAgICAgICAgICBidG4gPSBnci5CdXR0b24oIlNlbmQiLCB2YXJpYW50PSJwcmltYXJ5IikNCiAgICByZXR1cm4gc3RhdCwgaW5wLCBuLCBidG4NCg0Kd2l0aCBnci5CbG9ja3MoKSBhcyBkZW1vOg0KICAgIHN0YXQgPSBnci5TdGF0ZSh7fSkNCiAgICBzdGF0LnZhbHVlLnVwZGF0ZSh7InRhYnMiOiB7Im1haW4iOiAwLCAic3ViIjogMSwgImV4dHJhIjogMn19KQ0KICAgIHdpdGggZ3IuVGFicygpIGFzIHRhYnM6DQogICAgICAgIHdpdGggZ3IuVGFiSXRlbSgibWFpbiIsIGlkPTApOg0KICAgICAgICAgICAgc3RhdC52YWx1ZS51cGRhdGUoeyJpZCI6IDAsICJ0YWIiOiAibWFpbiJ9KQ0KICAgICAgICAgICAgc3RhdCwgaW5wLCBuLCBidG4gPSBpb19jb21wb25lbnRzKHN0YXQpDQogICAgICAgICAgICBidG4uY2xpY2soc2VuZF9hbmRfc3dpdGNoLCBbc3RhdCwgbl0sIFt0YWJzLCBpbnBdKQ0KICAgICAgICAgICAgZ3IuRXhhbXBsZXMoWydBJywgJ0InLCAnQyddLCBbaW5wXSkNCg0KICAgICAgICB3aXRoIGdyLlRhYkl0ZW0oInN1YiIsIGlkPTEpOg0KICAgICAgICAgICAgc3RhdC52YWx1ZS51cGRhdGUoeyJpZCI6IDEsICJ0YWIiOiAic3ViIn0pDQogICAgICAgICAgICBzdGF0LCBpbnAsIG4sIGJ0biA9IGlvX2NvbXBvbmVudHMoc3RhdCkNCiAgICAgICAgICAgIGJ0bi5jbGljayhzZW5kX2FuZF9zd2l0Y2gsIFtzdGF0LCBuXSwgW3RhYnMsIGlucF0pDQogICAgICAgICAgICBnci5FeGFtcGxlcyhbJ0EnLCAnQicsICdDJ10sIFtpbnBdKQ0KDQogICAgICAgIHdpdGggZ3IuVGFiSXRlbSgiZXh0cmEiLCBpZD0yKToNCiAgICAgICAgICAgIHN0YXQudmFsdWUudXBkYXRlKHsiaWQiOiAyLCAidGFiIjogImV4dHJhIn0pDQogICAgICAgICAgICBzdGF0LCBpbnAsIG4sIGJ0biA9IGlvX2NvbXBvbmVudHMoc3RhdCkNCg0KICAgICAgICAgICAgYnRuLmNsaWNrKHNlbmRfYW5kX3N3aXRjaCwgW3N0YXQsIG5dLCBbdGFicywgaW5wXSkNCiAgICAgICAgICAgIGdyLkV4YW1wbGVzKFsnQScsICdCJywgJ0MnXSwgW2lucF0pDQoNCmRlbW8ubGF1bmNoKHNob3dfZXJyb3I9VHJ1ZSk=)

**Comment by ZhengYuan-Public** [comment]
Since this thread is still open, I wonder what's the right method of doing this now?

I've wrote a function that can convert my input to docx/pdf file with pandoc. The function returns a `gr.DownloadButton` component which works. But when the updated button is pressed again, it shouldn't trigger the original `btn.click()` and run the function again.

I'm thinking of using a generate button and a download button and change their `visiable` property, but I can't find the right way to do it. The code is something like this:

```python
with gr.Column():
    btn_pdf_gen = gr.Button("Generate Markdown", size='md', visible=True)
    btn_pdf_dl = gr.DownloadButton(visible=False)

def handle_btn_pdf_click(summary_content, _btn_pdf_gen, _btn_pdf_dl):
    if _btn_pdf_gen.visible:
        # If btn_<format>_gen is visible, nothing is generated yet
        if not summary_content:
            gr.Warning("Summary Content is empty!")
        # generate_summary_pdf(summary_content) will return a gr.DownloadButton()
        return gr.Button(visible=False), generate_summary_pdf(summary_content)
    else:
        # Already generated, do nothing
        return _btn_pdf_gen, _btn_pdf_dl

btn_pdf_gen.click(
      fn=handle_btn_pdf_click,
      inputs=[summary_results, btn_pdf_gen, btn_pdf_dl],
      outputs=[btn_pdf_gen, btn_pdf_dl])
```

And apparently `_btn_pdf_gen.visible` is not available... I'm using `gradio 5.30.0` and can't find a workaround, any tips?@abidlabs 

Thanks ❤

**Comment by ZhengYuan-Public** [comment]
> Hi [@ZhengYuan-Public](https://github.com/ZhengYuan-Public) I'd suggest using a `gr.State` component to store the visibility and pass that as an input into your function. See [gradio.app/guides/state-in-blocks](https://www.gradio.app/guides/state-in-blocks)

It worked! Thanks :D

**Comment by by321** [comment]
For idea 1, not compatible with API is fine. Most code are not meant to be called via API, there's no need to cripple them for API accessibility.

Also, it would be very nice to be able to both read and write properties, for example change the selected item in a Gallery:

```
def callback_with_full_component(g):
  print(g.selected_index)
  g.selected_index=len(g.value)-1 #select the last item

```

**Comment by abidlabs** [comment]
@by321 you can already set the component properties by returning, e.g. `gr.Gallery(selected_index=1)` from your function

**Comment by by321** [comment]
> [@by321](https://github.com/by321) you can already set the component properties by returning, e.g. `gr.Gallery(selected_index=1)` from your function

Yes I know, but that's re-creating a component just to change one property, which is annoying and inconvenient. Also, you don't want component creation code littered all over the place, so it becomes necessary to put it in a separate function, for example:

```
def create_gallery(images,idx):
  return gr.Gallery(value=images, selected_index=idx, show_label=True, ...)

```
so if I want to change show_label to False later, or add or delete a parameter, I have only once place to change.

Yes I know, you can declare a separate GALLERY_SHOW_LABEL variable and use that all over the place. But that's not practical because you don't know what you want to change later, and you don't want a separate variable for every parameter you use.

**Comment by pngwn** [comment]
You can just pass the kwargs you want to update. You don't have to pass everything.

**Comment by by321** [comment]
OK this is news to me. 

If that's how it works, how do you know if the code is modifying an existing component or creating a new component and using default values for the unspecified keywords ? For example:

```
    gr.Image(interactive=False)
```
It should be creating a new component with default 'numpy' for the unspecified type. Suppose existing component's type is not 'numpy', how does it know to change only interactive but not type ?

**Comment by abidlabs** [comment]
Actually it's just updating the existing component! All unspecified properties are left unchanged from what they are prior to the update

**Comment by by321** [comment]
I mean "gr.Image(interactive=False)" always creates a new component, right ? How can Python know if you want to create a new one or update existing one ? It can't.

In the case of returning from gradio event handler, the gradio framework does the extra work of updating only the properties that you specified at creation time, correct ?

**Comment by abidlabs** [comment]
> In the case of returning from gradio event handler, the gradio framework does the extra work of updating only the properties that you specified at creation time, correct ?

yes exactly

**Comment by by321** [comment]
OK, then this unexpected behavior (return is not really a return) should be described in component documentation. For example, for Image it currently says:

> As output component: Expects a numpy.array, PIL.Image, or str or pathlib.Path filepath to an image which is displayed.

The following should be added:  or a gradio Image component, and in this case, the values of keyword arguments used to create it will be copied to the existing component.

**Comment by pngwn** [comment]
Yeah that could probably be clearer but is trying to communicate the required signature for the main value. 

If you could file an issue about the documentation we can try to make it clearer.

**Comment by by321** [comment]
#12055

**Comment by aliabid94** [comment]
> Idea 1:
> 
> * When a component is used as an input, we should pass in a custom class whose primary value is the value of the component, but which also has attributes that reflect the attributes of the property. For example:
> 
> ```py
> def get_tab_index(tabs):
>   return tabs.selected
> 
> with gr.Blocks() as demo:
>   with gr.Tabs() as tab:
>   ...
>   btn.click(get_tab_index, tabs, number)
> ```

The tricky part of this API would be to have variable `tabs` in the event listener refer to the value of the component, but `tabs.selected` be the prop value. For example, below:

```python
def add(x, y)
  print(x.maximum, y.maximum)
  return x + y

a = gr.Number()
b = gr.Number()
c = gr.Number()

b.submit(add, [a, b], c)
```

In `return x + y` we need these to refer to the values of x and y, but in `print(x.maximum, y.maximum)` we need to have these attributes present. It may be possible by extending the class type of each value and adding the attributes there. Will explore this, not 100% sure this exact syntax will be possible.

> However, the problem with this is is that it would not be compatible with using Gradio apps via an API.
> 
> Idea 2:
> 
> * You can pass in a component prop as an input. For example:
> 
> ```py
> with gr.Blocks() as demo:
>   with gr.Tabs() as tab:
>   ...
>   btn.click(lambda x:x, tabs.selected, number)
> ```

Similar problem in this API, `tabs.selected` would normally compute to some number, but within this event listener definition, we need to instead return a reference to the component and the prop value, e.g. the tuple `(tabs, "selected")`. May be able to overwrite `__getattr__` for components to enable this. Again still exploring if this exact syntax is possible.

Want to suggest another API that is quite different from everything we've seen in Gradio so far, but I think will save a lot of headache for users.

```python
num = gr.Number()

increase_btn = gr.Button("Increase")
increase_maximum_btn = gr.Button("Increase Maximum")
set_to_maximum_btn = gr.Button("Set to Maximum")

def increase()
  num.value += 1

increase_btn.click(increase)

def increase_maximum():
  num.maximum += 1

increase_maximum_btn.click(increase_maximum)

def set_to_maximum()
  num.value = num.maximum

set_to_maximum_btn.click(set_to_maximum)
```

This is an extremely different API that takes no inputs or outputs, and dynamically makes calls to the frontend (perhaps via websocket instead of SSE) to collect and set variable values. This does not allow API calls via a client in the traditional sense, but there is an assumption that any event listener that involves significant access to component properties is probably not meant for client calls, and requires frontend API use. 

However the benefits:
- Solves the issue of having to keep track of many different inputs/outputs
-  No need to redeclare variables, saving lots of code
- Allows intuitive python access to get and set component values and props.

Thoughts?

**Comment by abidlabs** [comment]
We discussed this before a bit here: https://github.com/gradio-app/gradio/issues/4926#issuecomment-1637132250, personally it seems like we're giving up a lot (API/MCP usage) if users start using this approach or mixing it up. What about passing in the component itself instead of your value if you type hint a parameter with a component, e.g.:

```py
def increase_maximum(num: gr.Number):
  num.maximum += 1
  return num

increase_maximum_btn.click(increase_maximum, number, number)
```

seems more inline with our existing usage. Such functions can immediately be inspected and removed from api/mcp just based on their signatures instead of having to inspect the function body, etc.

**Comment by freddyaboulton** [comment]
Was thinking something similar in my synch with Ali hehe!

**Comment by abidlabs** [comment]
Yay finally supported via https://github.com/gradio-app/gradio/pull/12267! (Follow up PRs will add support for more props)
